### PR TITLE
[Networking] Updating to `2019-09-01` / fixing `azurerm_virtual_hub` / VPN Resources

### PR DESCRIPTION
**Pull request by tombuildsstuff**
We ultimately ended up pulling the `azurerm_virtual_hub` Data Source & Resource from v1.37 since the resource design didn't match how the API worked. As such this PR reintroduces the `azurerm_virtual_hub` resource by introducing the dependencies necessary to use it.

Whilst it's unfortunate this PR is so large - due to the fact that this is already merged it makes sense to bundle these together - however traditionally we'd split these apart, in particular the SDK upgrade.

This PR includes:

* updating the `network` API to `2019-09-01`
* re-introducing the `azurerm_virtual_hub` Data Source / Resources:
    - removing the `express_route_gateway_id` field from the Data Source & Resource
    - removing the `p2s_vpn_gateway_id` field from the Data Source & Resource
    - removing the `s2s_vpn_gateway_id` field from the Data Source & Resource
    - removing the `route` block from the Data Source
    - removing the `virtual_network_connection` block from the Data Source & Resource
* removing the `security_provider_name` field from the `azurerm_virtual_wan` resource since this is no longer available in the API
* New Resource: `azurerm_point_to_site_vpn_gateway`
* New Resource: `azurerm_vpn_gateway`
* New Resource: `azurerm_vpn_server_configuration`

---

## Acceptance Test Results:

Data Source: `azurerm_virtual_hub`:

Screenshot 2019-11-27 at 16 15 49

Data Source: `azurerm_virtual_wan`:

Screenshot 2019-11-27 at 11 47 54

Resource: `azurerm_point_to_site_vpn_gateway`:

Screenshot 2019-11-28 at 17 44 15

Resource: `azurerm_virtual_hub`:

Screenshot 2019-11-27 at 11 47 33

Resource: `azurerm_virtual_hub_connection`:

Screenshot 2019-11-28 at 14 23 24

Resource: `azurerm_vpn_gateway`:

Screenshot 2019-11-28 at 14 23 24

Resource: `azurerm_vpn_server_configuration`:

Screenshot 2019-11-27 at 11 47 06

Resource: `azurerm_virtual_wan`:

Screenshot 2019-11-27 at 12 10 53

### CODE REVIEW

File: azurerm/resource_arm_virtual_hub.go [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/features"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/locks"
 	aznetwork "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/network"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
 )
 
+var virtualHubResourceName = "azurerm_virtual_hub"
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_hub.go]: Should this be a constant?

File: azurerm/resource_arm_virtual_hub.go [hunk 2]
```diff
@@ -167,36 +122,38 @@ func resourceArmVirtualHubCreateUpdate(d *schema.ResourceData, meta interface{})
...
 		Tags: tags.Expand(t),
 	}
 
-	if v, ok := d.GetOk("s2s_vpn_gateway_id"); ok {
-		s2sVpnGatewayId := v.(string)
-		parameters.VirtualHubProperties.VpnGateway = &network.SubResource{
-			ID: &s2sVpnGatewayId,
-		}
+	// todo: confirm if this is needed
+	// whilst there's a future here, the API is broken
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_hub.go]: Should we open an issue on the SDK and link back here?

File: azurerm/resource_arm_point_to_site_vpn_gateway_test.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+	return nil
+}
+
+func testAccAzureRMAzureRMPointToSiteVPNGateway_basic(rInt int, location string) string {
+	template := testAccAzureRMAzureRMPointToSiteVPNGateway_template(rInt, location)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_point_to_site_vpn_gateway" "test" {
+  name                        = "acctestp2svpng-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_point_to_site_vpn_gateway_test.go]: Should we have some capitals here to check casing?

File: azurerm/resource_arm_virtual_hub_test.go [hunk 4]
```diff
@@ -193,26 +210,18 @@ func testCheckAzureRMVirtualHubDestroy(s *terraform.State) error {
...
-  resource_group_name = "${azurerm_resource_group.test.name}"
-  location            = "${azurerm_resource_group.test.location}"
-}
+%s
 
 resource "azurerm_virtual_hub" "test" {
-  name                = "acctest-VirtualHub-%d"
-  resource_group_name = "${azurerm_resource_group.test.name}"
-  location            = "${azurerm_resource_group.test.location}"
+  name                = "acctestvhub-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_virtual_hub_test.go]: Should we have some capitals here to check casing?

File: azurerm/resource_arm_vpn_gateway.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"location": azure.SchemaLocation(),
+
+			"virtual_hub_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_vpn_gateway.go]: Should we valdiate this ID?

File: azurerm/resource_arm_vpn_gateway.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+						},
+					},
+				},
+			},
+
+			"scale_unit": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_vpn_gateway.go]: Validate this with > 0?

File: azurerm/resource_arm_vpn_gateway_test.go [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+	return nil
+}
+
+func testAccAzureRMVPNGateway_basic(rInt int, location string) string {
+	template := testAccAzureRMVPNGateway_template(rInt, location)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_vpn_gateway" "test" {
+  name                = "acctestvpng-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_vpn_gateway_test.go]: Should we have some capitals here to check casing?

File: azurerm/resource_arm_vpn_server_configuration.go [hunk 8]
```diff
@@ -0,0 +1,832 @@
...
+			})
+		}
+	}
+
+	radiusAddress := ""
+	if input.RadiusServerAddress != nil {
+		radiusAddress = *input.RadiusServerAddress
+	}
+
+	// TODO: confirm if secret is returned or if we need to look it up
```
Comment (katbyte) [inline azurerm/resource_arm_vpn_server_configuration.go]: Still a todo?

File: azurerm/resource_arm_vpn_server_configuration_test.go [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+	return nil
+}
+
+func testAccAzureRMAzureRMVPNServerConfiguration_azureAD(rInt int, location string) string {
+	template := testAccAzureRMAzureRMVPNServerConfiguration_template(rInt, location)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_vpn_server_configuration" "test" {
+  name                    = "acctestrg-%d"
```
Comment (katbyte) [inline azurerm/resource_arm_vpn_server_configuration_test.go]: Should this be:
```suggestion
  name                    = "acctest-VPNCFG-%d"
```

File: azurerm/resource_arm_vpn_server_configuration.go (hunk 8 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_vpn_server_configuration.go]: this can be removed

**Comment by tombuildsstuff** [comment]
Acceptance Tests still pass after PR updates 👍

File: azurerm/resource_arm_virtual_hub.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_virtual_hub.go]: after re-testing turns out this isn't needed ¯\_(ツ)_/¯

**Comment by neil-yechenwei** [comment]
Hi @tombuildsstuff , thanks for this PR. However, I have a question about this. I remember you wanna make the property "virtual_network_connection" as separate resource. But in this PR, I didn't find the separate resource for this property. Seems that it has been removed from the resource. Maybe I missed something. Can I know the reason why this property should be removed?

**Comment by tombuildsstuff** [comment]
@neil-yechenwei it turns out the API doesn't work so we had to pull it

**Comment by neil-yechenwei** [comment]
> @neil-yechenwei it turns out the API doesn't work so we had to pull it

It's weird. Because I tried this property "virtual_network_connection" in version "2019-07-01" and I can add/edit virtual network connection successfully. So maybe it's a bug in version "2019-09-01", right?

**Pull request merged**